Add Integration for Marten

jabellard opened the pull request:
Integration for Marten

Generally, the features from the `HotChocolate.Data` package should work with any LINQ provider from which some `IQueryable<T>` can be retrieved. However, this is not the case with [Marten](https://martendb.io/). Pagination and projections work out of the box as expected, but filtering and sorting do not. LINQ expressions generated for filtering and sorting must first be translated in a format that is digestible for the Marten LINQ provider before they are applied to the `IQueryable<T>` object. This integration provides custom configurations to seamlessly integrate Marten with the `HotChocolate.Data` package.

Closes #5282

jabellard commented:
@michaelstaib , Given that this is my first contribution, it seems like I'll have to sign a CLA. I get that from you?

michaelstaib commented:
@jabellard yes ... I will send it to you to sign on slack.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ChilliCream/hotchocolate?pullRequest=5301) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ChilliCream/hotchocolate?pullRequest=5301) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ChilliCream/hotchocolate?pullRequest=5301) it.</sub>

michaelstaib commented:
@jabellard we have now everything setup for the CLA, click on the link in the CLA bot message and sign the agreement.

jabellard commented:
@michaelstaib Ok. Just signed.

jabellard commented:
Also I just noticed that The `CodeQl` check is failing after merging in the `main` branch: 
```
Error: /home/runner/work/hotchocolate/hotchocolate/src/StrawberryShake/SourceGenerator/test/CodeGeneration.CSharp.Analyzers.Tests/StarWars/Test.cs(9,18): error CS1061: 'IServiceCollection' does not contain a definition for 'AddStarWarsClient' and no accessible extension method 'AddStarWarsClient' accepting a first argument of type 'IServiceCollection' could be found (are you missing a using directive or an assembly reference?) [/home/runner/work/hotchocolate/hotchocolate/src/StrawberryShake/SourceGenerator/test/CodeGeneration.CSharp.Analyzers.Tests/StrawberryShake.CodeGeneration.CSharp.Analyzers.Tests.csproj]
```
This is not related to the changes in this PR, so I'm not sure what's going on there.

michaelstaib commented:
The issue with CodeQL has to do with source generators... will have a look in a separate PR

jabellard commented:
@michaelstaib , can you take a look? I want to work on finishing this up.

michaelstaib commented:
Its waiting for a review by @PascalSenn

PascalSenn commented:
@jabellard  Sorry i didnt complete the review so the comments were still pending 🤦 
I have a couple of questions.

jabellard commented on src/HotChocolate/Marten/src/Data/MartenExpressionTranslator.cs:
```diff
@@ -0,0 +1,41 @@
+
+    private static bool ReferenceComparisonExpressionShouldBeRemoved(BinaryExpression expression)
+    {
+        return expression.NodeType == ExpressionType.NotEqual
```
Yes, but only if it's on the base object of the query, and not a member access expression.
So:
`where([x] != null AndAlso [x].Rating > 50)` will get simplified to `where([x].Rating > 50)`
But:
`where([x].NestedObject  != null AndAlso [x].NestedObject.Rating > 50)` will get not get simplified.

jabellard replied:
Also, I don't rewrite any of the expressions. For a binary expression like `where([x] != null AndAlso [x].Rating > 50)`, to simplify, I simply ignore the left operand and return the right one.

jabellard replied:
What exactly does `context.InMemory` indicate? And no. Actually Marten will work just fine without the null checks, so we don't really need them.

PascalSenn replied:
@jabellard `context.InMemory` removes all the null checks. We do this for ef core to optimize the expressions. Do you think this could also work for marten?

jabellard replied:
@PascalSenn Ok. I see. Unlike EF, Marten does not support an in memory database, but `context.InMemory` evaluates to `true` when working with the Marten queryable for some reason so the null checks are added. However, given that Marten will work fine without the null checks, everything should work like a charm if we just get rid of the null checks at the following sites:
https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Data/src/Data/Filters/Expressions/Handlers/QueryableDefaultFieldHandler.cs#L112
https://github.com/ChilliCream/hotchocolate/blob/main/src/HotChocolate/Data/src/Data/Sorting/Expressions/Handlers/QueryableDefaultSortFieldHandler.cs#L73

What do you recommend as the best, least intrusive way to go about doing that for Marten?

jabellard replied:
@PascalSenn , Never mind. Figured it out. Can you take a look?

jabellard commented:
@PascalSenn Responded to comments. Just let me know if you have any other questions.

PascalSenn commented on src/HotChocolate/Marten/src/Data/Filtering/Extensions/RequestExecutorBuilderExtensions.cs:
```diff
@@ -0,0 +1,26 @@
+
+public static class RequestExecutorBuilderExtensions
+{
+    public static IRequestExecutorBuilder AddMartenFiltering(this IRequestExecutorBuilder requestExecutorBuilder)
```
@jabellard  Does marten have a native api too? if so we might want to change this to `AddMartenQueryableFiltering`. Because then we would still be able to add the native filtering stuff too.
Also wonder what @michaelstaib  thinks about it. specfic or less specific

jabellard replied:
@PascalSenn no. There's no other api.

PascalSenn commented:
@jabellard 

So i have changed the implementation to use InMemory and the default QueryableProvider. With this implementation we are now also able to use marten with resolver that return "IEnumerable" or just IQueryable without breaking them.

I have also added tests for sorting and filtering. We have a set of regression tests that we run for each provider.

I fixed already a bunch of issues but i see three main points that are not working with the Marten SQL transaltor:
1. `in` Expressions  like `where: {foo: {bar: { in: [1,2,3]}}}`  (i think i know what this issue is)
2. custom expressions (new feature in v13) e.g. `x => x.FirstName + " " + x.LastName`

Can you have a look with the marten team if we can make these expressions work (or otherwise what a alternative would be):
```
"_s0 => (_s0.Bars.Count == 10)"
```
```
Marten Command Failure:$\nWITH mt_temp_id_list3CTE as (\nselect ctid, unnest(CAST(ARRAY(SELECT jsonb_array_elements(CAST(d.data -\u003E\u003E \u0027Bars\u0027 as jsonb))) as jsonb[])) as data from public.mt_doc_queryablefiltervisitorexpressiontests_foo as d\n)\n  , mt_temp_id_list4CTE as (\nselect ctid, count(*) as data from mt_temp_id_list3CTE as d group by ctid having count(*) =:p0\n)\n select d.id, d.data from public.mt_doc_queryablefiltervisitorexpressiontests_foo as d where d.ctid in (select ctid from mt_temp_id_list4CTE)$\n$\n22023: cannot extract elements from an object
```

PascalSenn commented:
Also a major issues are the List operations. Most of them seem not to work?

PascalSenn commented:
@michaelstaib  
It seems like marten does not support most of the list operations `all` `any` `some` and also has limited support for custom expressions. 

My suggestion would be deactivating list filtering for the marten queryable provider and remove the tests for the custom expressions (as it is more a db queryable provider limitation rather than a hc limitation)

jabellard commented:
> @michaelstaib It seems like marten does not support most of the list operations `all` `any` `some` and also has limited support for custom expressions.
> 
> My suggestion would be deactivating list filtering for the marten queryable provider and remove the tests for the custom expressions (as it is more a db queryable provider limitation rather than a hc limitation)

@PascalSenn : I think you have been looking at an old version of the Marten docs. At the moment, Marten does not have support for the `all` quantifier operation, but does have support for the `any` and `some` operations: https://martendb.io/documents/querying/linq/child-collections.html#quantifier-operations-within-child-collections

PascalSenn commented:
well it seems like the expressions that we compile do not work with marten at least ;) there are failing tests in the repo, you can check it out and run the tests. maybe you have an idea what goes wrong

oskardudycz commented:
Hey folks, thanks for the effort on making the initial Marten+HotChocolate integration. I just skimmed the discussion and PR, I plan to have a wider look soon. A few comments from my side:

>It seems like most of the list operations will not work:
> https://martendb.io/v3/documentation/documents/querying/linq

Those docs are for v3; we have the right now v5, which had a big revamp around Linq. Check the docs at https://martendb.io/documents/querying/linq/operators.html for the latest version. Still, even for the v3 we had pretty good support for the most common scenarios. Could you expand on what's not working or link me to the failing scenarios? Of course, we cannot support all the possible permutation of scenarios as an in-memory provider, but still, I think that majority of the common should work.

Regarding returning to memory and then making filtering, I think that's not the best approach. It'd be similar to the initial EF Core release, which just simulated the queries and was not performant. My recommendation would be to try to see what we can do to make the scenarios work, and if some don't, then list them and fail with the specific error message, the user will get at least an understanding of the limitations.

Thoughts?

michaelstaib commented:
I also would like to avoid doing in-memory filtering as it could be very heavy on memory and yield suboptimal performance. Lets, try to get as many things in as possible and just remove the things that do not work from the provider in HC.

PascalSenn commented:
Hi @oskardudycz 

So we do project to expressions that then are interpreted by marten. So on the HotChocolate side we do not really filter in memory whenever possible.

So the `List.Contains` stuff is already fixed, @jabellard  is now looking into how to solve the issues that we have with the operations `all` `some` `none` and `any`

- `any` This is a true/false value that indicates if the collection is empty. I think this can be easy worked around with `foo.Bars.Count() >0`
- `some` matchs am element of the collection when one of the elements of the nested collection matchs the condition. So in expression terms `foos.Where(x => x.Bars.Any(x => x.Baz == 1 && x.Qux == 2)`
- `none` is the negation of `some`.  (so should be easy once `some` is solved)
- `all` matchs an element of the collection when all of the elements of the nested collection match the condition. So in expression terms `foos.Where(x => x.Bars.All(x => x.Baz == 1 && x.Qux == 2)`

jabellard commented:
Yes!
@oskardudycz , as @PascalSenn  mentioned, filtering as as well as the other features such as sorting provided by Hot Chocolate will execute queries natively in the database as opposed to in memory. The references to `InMemory` in the thread, if I remember correctly, are about how Hot Chocolate will generate specialized expressions that, for instance, will include null guards when in memory LINQ provider is used. 

Regarding the list operators. I anticipate that all of them, except the `all` operator, which Marten does not support at the moment, will work. I got them working in a project that I have. The tests that @PascalSenn added are not working yet, but from what I can tell so far, this is due to an issue with how the documents saved to the database during the test session are serialized, hence the wierd error:
```
Marten Command Failure:$\nWITH mt_temp_id_list3CTE as (\nselect ctid, unnest(CAST(ARRAY(SELECT jsonb_array_elements(CAST(d.data -\u003E\u003E \u0027Bars\u0027 as jsonb))) as jsonb[])) as data from public.mt_doc_queryablefiltervisitorexpressiontests_foo as d\n)\n , mt_temp_id_list4CTE as (\nselect ctid, count(*) as data from mt_temp_id_list3CTE as d group by ctid having count(*) =:p0\n)\n select d.id, d.data from public.mt_doc_queryablefiltervisitorexpressiontests_foo as d where d.ctid in (select ctid from mt_temp_id_list4CTE)$\n$\n22023: cannot extract elements from an object
```

 I should get to the bottom of why that's happening very soon and fix those tests.

As far as the `all` operator is concerned, for now, I think we should just keep it disabled until support for it is added to Marten.

jabellard commented:
@PascalSenn, I fixed the issue with serialization the test data. There's one more thing that we need to figure out.
A query like :
```
{
                        root(where: {
                            fooNested: {
                                some: {
                                    bar: {
                                        eq: "a"
                                    }
                                }
                            }
                        }){
                            fooNested {
                                bar
                            }
                        }
                    }
```
Should work like a charm. I know because I've tested similar queries in a project that I have. However, the test for it is not passing. This is because the sql generated in the test environment is malformed. The generated sql for the query above should be:
```
WITH mt_temp_id_list1CTE as (
select ctid, unnest(CAST(ARRAY(SELECT jsonb_array_elements(CAST(d.data ->> 'FooNested' as jsonb))) as jsonb[])) as data from public.mt_doc_queryablefiltervisitorlisttests_foo as d
 )
   , mt_temp_id_list2CTE as (
 select ctid, data from mt_temp_id_list1CTE as d where d.data ->> 'Bar' = 'a'
 )
  select d.id, d.data from public.mt_doc_queryablefiltervisitorlisttests_foo as d where d.ctid in (select ctid from mt_temp_id_list2CTE)
```
However, what is generated in the test environment is (with added comments):
```
.param set @__p_0 'a'

--- not needed============================================================
SELECT "d"."Id"
FROM "Data" AS "d"
WHERE EXISTS (
    SELECT 1
    FROM "FooNested" AS "f"


-- what is actually needed============================================================
 WITH mt_temp_id_list1CTE as (
 select ctid, unnest(CAST(ARRAY(SELECT jsonb_array_elements(CAST(d.data ->> 'FooNested' as jsonb))) as jsonb[])) as data from public.mt_doc_queryablefiltervisitorlisttests_foo as d
 )
   , mt_temp_id_list2CTE as (
 select ctid, data from mt_temp_id_list1CTE as d where d.data ->> 'Bar' = :p0
 )
  select d.id, d.data from public.mt_doc_queryablefiltervisitorlisttests_foo as d where d.ctid in (select ctid from mt_temp_id_list2CTE)



 --- not needed============================================================
    WHERE "d"."Id" = "f"."FooId" AND "f"."Bar" = @__p_0)
```

The part in the middle is what we need.  We don't need the parts at the top and bottom. Can we fix this by configuring the test setup differently?

jabellard commented:
@PascalSenn : Never mind. I got all the list tests working. I just had to get rid of the old snapshot files to get things working.

jabellard commented:
All tests are filtering are passing.

jabellard commented:
> All tests for filtering are passing if I run each test class in that project separately. However, if I try running all the tests in that project at once, then I get some snapshot mismatch errors.

I take that back. The in/nin tests are failing which is weird, since I fixed them earlier.

jabellard commented:
I fixed the issue with the test setup. All tests are now passing except for tests regarding:
- Nullable enum values
- Custom expressions

jabellard commented:
@PascalSenn , @oskardudycz : We're  super close to finishing up this integration.

There's been a few outstanding issues holding up this PR:

- Invoking custom handlers that @PascalSenn implemented as opposed to the  default ones in the test environment:
     I reconfigured the test setup, and resolved that.

- Failing tests for the `all` list operator
    At the moment, Marten has no support for that operator, so I removed those tests. @PascalSenn , is there a way to disable 
    that operation when the Marten filtering provider is used?

- Failing tests for custom expressions
   That's expected. Marten can't support every possible permutation of building custom expressions, so I removed the tests for the ones that are not supported.

- Failing tests for `in` and `nin` queries for nullable enum values
     This is now the only outstanding issue that's holding off this PR. The tests are failing because of a limitattion in Marten for parsing expressions used to filter nullable enum values. I resolved that issue, and submitted a PR for it: https://github.com/JasperFx/weasel/pull/70 .  @oskardudycz , Can you take a look? I've already tested this locally, and got all the tests passing. One this gets merged in, we can just reference the latest version of Marten in this project to pass the outstanding tests.

PascalSenn commented:
@oskardudycz : Do you have any suggestion how we could get `all` to work with marten?

oskardudycz commented:
Folks, I apologise for the lack of feedback. I was (and still am) a bit too busy last week. I'll try to investigate that the next week.

jabellard commented:
@PascalSenn , @oskardudycz : I took a look into implementing the `All` operation for Marten, and have a very good idea on how to go about doing it. I'll give it a shot this weekend.

jabellard commented:
@PascalSenn , @oskardudycz : I added support to Marten for the `all` operation  https://github.com/JasperFx/marten/pull/2368  and added back the corresponding tests.

Once that PR, along with https://github.com/JasperFx/weasel/pull/70, get merged in, I'll just reference the latest version of Marten in this project to pass the outstanding tests, and we'll be good to go!

michaelstaib commented:
Can we get this wrapped up until the end of November so we can ship it with 13.0.0?

michaelstaib commented:
No worry if not :) we have 13.1 planned for end of January ... more for my internal planning.

jabellard commented:
> Can we get this wrapped up until the end of November so we can ship it with 13.0.0?

@michaelstaib : You have a date in mind for the 13.0.0 release?

michaelstaib commented:
@jabellard yes ... first RC build at the end of this month ... but if we cannot make it we just put it in 13.1 which should be out end of January.

jabellard commented:
@michaelstaib , @PascalSenn : I was able to reference the latest version of Marten to pass the outstanding tests. All the tests are now passing, so I think this should be ready to go.

michaelstaib commented:
I will have look tonight. Are you on slack?

jabellard commented:
> I will have look tonight. Are you on slack?

@michaelstaib , Yeah. I'm on there.

michaelstaib commented:
This looks good ... I am almost done with the review .... it will go in today.

oskardudycz commented:
Hey Guys, just wanted to say that I appreciate a lot your work on getting that done and delivering this feature. Kudos to all involved 💪🙂
